test/unit/ngsiv2/httpBindings-test.js

```diff
@@ -0,0 +1,738 @@
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('When a request arrives to the IoT Agent without Content-type header', function() {
+        var postOptions = {
+            url: 'http://localhost:' + config.http.port + '/iot/d',
+            method: 'POST',
+            qs: {
+                i: 'urn:x-iot:smartsantander:u7jcfa:fixed:t311',
+                k: '1234',
+                t: '2016-05-11T10:12:26.476659Z'
+            },
+            body: 'bat|75.0'
+        },
+        provisionOptions = {
+            url: 'http://localhost:' + config.iota.server.port + '/iot/devices',
+            method: 'POST',
+            json: utils.readExampleFile('./test/deviceProvisioning/provisionDeviceProduction.json'),
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            }
+        };
+
+        beforeEach(function(done) {
+            nock.cleanAll();
+            // Note: /v1/updateContext response is not processed by IOTA so its content is irrelevant,
+            // as far as it is a 200 OK
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+            .post('/v1/updateContext')
+            .reply(200 , '{}');
+
+            contextBrokerMock
+            .post('/v2/entities/urn:x-iot:smartsantander:u7jcfa:fixed:t311/attrs')
+            .reply(204);
+
+            request(provisionOptions, function(error, response, body) {
+                done();
+            });
+        });
+
+        it('should end up with a 200OK status code', function(done) {
+            request(postOptions, function(error, response, body) {
+                should.not.exist(error);
+                response.statusCode.should.equal(200);
+
+                done();
+            });
+        });
+    });
+
+    describe('When a measure arrives to the IoTA for a device belonging to a configuration', function() {
+        var getOptions = {
+            url: 'http://localhost:' + config.http.port + '/iot/d',
+            method: 'POST',
+            qs: {
+                i: 'MQTT_2',
+                k: '80K09H324HV8732',
+                d: 'c|23'
+            }
+        },
+        groupCreation = {
+            url: 'http://localhost:4041/iot/services',
+            method: 'POST',
+            json: utils.readExampleFile('./test/groupProvisioning/provisionAliasGroup.json'),
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            }
+        };
+
+        beforeEach(function(done) {
+            contextBrokerMock
+            .matchHeader('fiware-service', 'smartGondor')
+            .matchHeader('fiware-servicepath', '/gardens')
+            .post('/v2/entities/Second%20UL%20Device/attrs',
+                utils.readExampleFile('./test/unit/ngsiv2/contextRequests/unprovisionedAliasMeasure.json'))
```

Comment:
This is:

```
{
   "Correlation": {
    	"type": "Integer",
    	"value": "23"
    }
}
```

However, looking to test/groupProvisioning/provisionAliasGroup.json, the attribute is defined as an integer:

```
      "attributes": [
        {
          "object_id": "c",
          "name": "Correlation",
          "type": "Integer"
        }
]
```

So value should be `23` instead of `"23"`, isn't it?

Reply:
This is related to https://github.com/telefonicaid/iotagent-node-lib#attribute-alias-plugin-attributealias. In https://github.com/fermenreq/iotagent-ul/blob/ActiveAttsNgsi2/test/unit/ngsiv2/groupProvisioning/provisionAliasGroup.json, a group is provisioned and we define an attribute using an id (aka, *object_id*) together with the name:

```
{
  "services": [
    {
      "resource": "/iot/d",
      "apikey": "80K09H324HV8732",
      "entity_type": "SensorMachine",
      "cbHost": "http://unexistentHost:1026",
      "commands": [],
      "lazy": [],
      "attributes": [
        {  
          "object_id": "c",
          "name": "Correlation",
          "type": "Integer"
        }
      ]
    }
  ]
}
```

As the IOTA is agnostic of attribute alias functionalities, when a new observation arrives for this attribute, it calls to iotAgentLib.update function with the following arguments:
- device.name: Second UL Device
- device.type: AnMQTTDevice
- values: [{"name":"c","type":"string","value":"23"}]

This is correct because, as I said, the IOTA does not know anything about alias and its the role of the plugin to do all the required transformations.

The current version of the attribute alias does the following: in https://github.com/dcalvoalonso/iotagent-node-lib/blob/task/activeAttsNgsi2/lib/plugins/attributeAlias.jsL87, it extracts all mappings (alias) for the device's attributes and then it applies them, which implies the replacement of the attribute name and type in https://github.com/dcalvoalonso/iotagent-node-lib/blob/task/activeAttsNgsi2/lib/plugins/attributeAlias.js#L70.

As you can see, nothing is done for the attribute's value so I guess that in order to have a correct functionality, we should add something along this address: for NGSIv2, to parse the attribute's value considering the type as defined by the alias/mapping.

Reply:
I would like to know your point of view @fermenreq and @fgalan! ;)

Reply:
To check if I'm understanding correctly...

The value of the attribute is send as "23" instead of 23 to CB due to alias plugin avoids it. That, is when the long name for the attribute ("Correlation") is used in the UL measure instead of the short one ("c") then alias plugin is not activated and 23 is sent to CB (I undesrtand that @fermenreq last test is showing that).

Is my interpretation correct?

Reply:
The value of the attribute "Correlation" sent is 23. 

```
{
   "Correlation": {
    	"type": "Integer",
    	"value": 23
    }
}
```
It would be nice to develop @dcalvoalonso answer

Reply:
@fermenreq you mean that't the value when `Correlation|23`, isn't it?

Please note I'm asking a wider question:

> The value of the attribute is send as "23" instead of 23 to CB due to alias plugin avoids it. That, is when the long name for the attribute ("Correlation") is used in the UL measure instead of the short one ("c") then alias plugin is not activated and 23 is sent to CB (I undesrtand that @fermenreq last test is showing that).

@dcalvoalonso could you confirm that's the behaviour? Thx!

Reply:
@fgalan you are more than right. If the observation is sent to the IOTA-UL with 'Correlation|23' everything works because we are not using an alias. However, the use of the short one ("c") is related to the alias plugin, which does not implement type casting for attributes' values.

Reply:
Thanks. Now I think I understand what the situation is.

It would be great that type casting also works for alias. If it is easy, I'd suggest to go for it. If it is complex, an issue about the lack of functionality should be opened and warned in the documentation (the documentation I mention in my other comment: https://github.com/telefonicaid/iotagent-ul/pull/242#discussion_r170913874)).

Reply:
I have open a new issue to document this requirement: https://github.com/telefonicaid/iotagent-node-lib/issues/583

Let's see if we can do it! ;)

Reply:
If the alias pluging is not going to be fixed before merging this PR, then a suitable FIXME mark should be added to this test (and maybe others in the same situation), e.g:

```
// FIXME : casting to numbers in not supported when using alias, see issue https://github.com/telefonicaid/iotagent-node-lib/issues/583
// Once that issue gets solved, the .json used by this test should be adjusted
```

Reply:
I have done a first implementation of this issue in https://github.com/dcalvoalonso/iotagent-node-lib/pull/7. I think that it is quite simple and we can solve the PR quickly so this should be ready for solving this PR.